cpp/src/Ice/ConnectionFactory.cpp

```diff
@@ -980,8 +980,18 @@ IceInternal::OutgoingConnectionFactory::ConnectCallback::nextEndpoint()
 {
     try
     {
+        auto self = shared_from_this();
         assert(_endpointsIter != _endpoints.end());
-        (*_endpointsIter)->connectors_async(_selType, shared_from_this());
+        (*_endpointsIter)->connectorsAsync(
```

Comment:
With this code, you make the assumption that the OutgoingConnectionFactory is not destroyed when the callback executes.

If somehow the logic of the code makes this guarantee, we should add a comment:
```
This OutgoingConnectionFactory is necessarily still around when the callbacks execute because ...
```

Otherwise, you should use our usual pattern:
```
auto self = shared_from_this();
```

and then capture self in the lambdas.

Reply:
For the common TCP path, "this" is correct because:
 - Instance hold onto this _outgoingConnectionFactory
 - the callbacks are executed by the endpointHostResolver thread
 - we join this thread in Instance::destroy() before setting _outgoingConnectionFactory to nullptr;